### PR TITLE
S2U-9 5.1.1.5 Tests & Quizzes: PDF export for assessment submissions

### PR DESCRIPTION
**Pull request by JuanDavid102**

### CODE REVIEW

**Comment by mpellicer** [comment]
Let's leave this open until next Tuesday so the community can make comments. @frasese would you mind doing an initial review?

**Comment by JuanDavid102** [comment]
Ok

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java [hunk 1]
```diff
@@ -0,0 +1,681 @@
...
+		public ImageMapQuestionCellEvent(ArrayList<Circle> answerCircles, ArrayList<Rectangle> answerRectangles, float originalWidth, float originalHeight) {
+			this.answerRectangles = answerRectangles;
+			this.answerCircles = answerCircles;
+			this.originalWidth = originalWidth;
+			this.originalHeight = originalHeight;
+		}
+
+        public void cellLayout(PdfPCell cell, Rectangle position, PdfContentByte[] canvases) {
+			PdfContentByte canvas = canvases[PdfPTable.TEXTCANVAS];
+            float x = position.getLeft();
```
Comment (jesusmmp) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:562]: Spacing

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+					}
+				}
+			}
+			document.close();
+
+			outputStream.flush();
+			outputStream.close();
+			faces.responseComplete();
+		} catch (Exception ex) {
+			ex.printStackTrace();
```
Comment (jesusmmp) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:394]: log.error

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java [hunk 3]
```diff
@@ -0,0 +1,681 @@
...
+				for (ItemContentsBean item : items) {
+					Long questionType = item.getItemData().getTypeId();
+					PdfPTable questionTable = new PdfPTable(2);
+					questionTable.setWidthPercentage(50f);
+					questionTable.setHorizontalAlignment(PdfPTable.ALIGN_LEFT);
+					this.addCellToTable(questionTable, ( rb.getFormattedMessage("current_question", new String[]{String.valueOf(++itemsIndex), String.valueOf(questionsCuantity)}) ), 3, 1);
+					this.addCellToTable(questionTable, (item.getPointsDisplayString() + "/" + item.getMaxPoints()), 3, 0);
+					document.add(new Paragraph(Chunk.NEWLINE));
+					document.add(questionTable);
+					if (questionType == TypeIfc.FILL_IN_NUMERIC || questionType == TypeIfc.CALCULATED_QUESTION) {
```
Comment (jesusmmp) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:171]: https://github.com/sakaiproject/sakai/pull/11809/files#diff-1f00a3cc1fa4cbdc1216b6ac893e483b958e7cb064707d7219f212e821902afeR171-R206

Refactor?

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java [hunk 4]
```diff
@@ -0,0 +1,681 @@
...
+  	}
+
+	public void addCellToTable(PdfPTable table, String content, int configuration, int color) {
+		PdfPCell cell;
+		switch (configuration) {
+            case 0:
+				table.setWidthPercentage(100f);
+				cell = new PdfPCell(new Paragraph(content, new Font(Font.TIMES_ROMAN, 16, Font.BOLD, Color.GRAY)));
+				cell.setBorderWidth(0);
+				table.addCell(cell);
```
Comment (jesusmmp) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:406]: table.addCell(cell); 
better out of cases (it is present on all case)

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java [hunk 5]
```diff
@@ -0,0 +1,681 @@
...
+		finalTable.addCell(finalQuestionCell);
+		
+		return finalTable;
+	}
+
+	public String cleanText(String text) {
+		int tableIndex = text.indexOf("<table");
+		String finalText = text;
+		while (tableIndex != -1) {
+			finalText = (text.substring(0, tableIndex) + text.substring(text.indexOf("</table>", tableIndex) + 8));
```
Comment (jesusmmp) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:675]: +8?. Avoid magic number

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java [hunk 6]
```diff
@@ -0,0 +1,681 @@
...
+					this.addCellToTable(shortSummaryTable, rb.getString("type." + item.getItemData().getTypeId()), 2, tableColor);
+					this.addCellToTable(shortSummaryTable, (!item.isUnanswered() || this.isAnswered(item) ? rb.getString("short_summary.answered.yes") : rb.getString("short_summary.answered.no")), 2, tableColor);
+					this.addCellToTable(shortSummaryTable, (item.getPointsDisplayString() + "/" + item.getMaxPoints()), 2, tableColor);
+				}
+			}
+
+			document.add(shortSummaryTable);
+ 			document.add(new Paragraph(Chunk.NEWLINE));
+			String comments = studentScoreBean.getComments();
+			if (comments != null && !StringUtils.equals(comments, "")) {
```
Comment (jesusmmp) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:148]: if (StringUtils.isNotEmpty(comments)) {

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java (hunk 3 above)
Comment (jesusmmp) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:171]: Code on lines above are similar except:  item.getFinArray() or item.getFibArray();
Maybe a new method could be created to put all these code only once. 

Something like:  

```
if (questionType == TypeIfc.FILL_IN_NUMERIC || questionType == TypeIfc.CALCULATED_QUESTION) {
  document = processQuestions(document, item.getFibArray())
} else if (questionType == TypeIfc.FILL_IN_BLANK) {
  document = processQuestions(document, item.getFinArray())
}
```

and the processQuestions will contain the same code

```
PdfPTable fillInNumTable = new PdfPTable(1);
fillInNumTable.setHorizontalAlignment(PdfPTable.ALIGN_LEFT);

int i = 0;
String questionText = "";
List<FinBean> fillNumericArray = item.getFinArray();
for (FinBean fillNumericItem : fillNumericArray) {
	if (i + 1 != fillNumericArray.size()) {
		questionText += fillNumericItem.getText() + "(" + (++i) + ")";
		int tableColor = (fillNumericItem.getIsCorrect()!= null && fillNumericItem.getIsCorrect()? 1 : (StringUtils.equals(fillNumericItem.getResponse(), "")? 2 : 3));
		this.addCellToTable(fillInNumTable, (i) + ". " + (StringUtils.equals(fillNumericItem.getResponse(), "")? rb.getString("no_answer.text") : fillNumericItem.getResponse() ), 4, tableColor);
	} else {
		questionText += fillNumericItem.getText();
	}
}
document.add(this.getQuestionTitle(questionText));
document.add(fillInNumTable);
```

It's only an idea...

Comment (JuanDavid102) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:171]: Ok. Thanks, i think it is a good idea, i will aply it

**Comment by mpellicer** [comment]
Please take care of the codacy issues.

**Review by mpellicer — CHANGES_REQUESTED** [review]
Changes requested.

File: samigo/samigo-app/pom.xml [hunk 7]
```diff
@@ -276,6 +276,16 @@
             <artifactId>jakarta.servlet.jsp.jstl-api</artifactId>
             <version>${sakai.jakarta.jstl-api.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.json</groupId>
+            <artifactId>json</artifactId>
+            <version>20210307</version>
```
Comment (mpellicer) [inline samigo/samigo-app/pom.xml:282]: Stick to the Sakai version, it is defined in Master, you probably don't need to include the version here.

**Comment by mpellicer** [comment]
Could you take a look @frasese  ? Thanks

File: samigo/samigo-app/pom.xml [hunk 8]
```diff
@@ -276,6 +276,16 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-imaging</artifactId>
```
[the remark this answers is not in the document]
Comment (JuanDavid102) [reply samigo/samigo-app/pom.xml:286]: I directly deleted. It is not used

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java [hunk 9]
```diff
@@ -0,0 +1,710 @@
+/**********************************************************************************
+ * $URL$
+ * $Id$
+ ***********************************************************************************
+ *
+ * Copyright (c) 2004, 2005, 2006, 2007, 2008, 2009, 2023 The Sakai Foundation
```
Comment (frasese) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:6]: Just 2023

**Comment by JuanDavid102** [comment]
[Report_Fadesta_uno mas segunda prueba-2.pdf](https://github.com/sakaiproject/sakai/files/12291163/Report_Fadesta_uno.mas.segunda.prueba-2.pdf)

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java [hunk 10]
```diff
@@ -0,0 +1,710 @@
...
+	 */
+	public void processAction(ActionEvent ae) {
+		try {
+			FacesContext faces = FacesContext.getCurrentInstance();
+			HttpServletResponse response = (HttpServletResponse) faces.getExternalContext().getResponse();
+			DeliveryBean deliveryBean = (DeliveryBean) ContextUtil.lookupBean("delivery");
+			StudentScoresBean studentScoreBean = (StudentScoresBean) ContextUtil.lookupBean("studentScores");
+
+			Document document = new Document(PageSize.A4, 20, 20, 20, 20);
+			ServletOutputStream outputStream = response.getOutputStream();
```
[the remark this answers is not in the document]
Comment (JuanDavid102) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:106]: Ok. what should i put in the finally block? Something like: log.error("Something fails"); or more?

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java [hunk 11]
```diff
@@ -0,0 +1,710 @@
...
+
+	/**
+	 * Method to create a cell using the sent text and configuration and added to the table sent
+	 * 
+	 * @param table - PdfPTable
+	 * @param content - String
+	 * @param configuration - int
+	 * @param color - int
+	 */
+	public void addCellToTable(PdfPTable table, String content, int configuration, int color) {
```
[the remark this answers is not in the document]
Comment (JuanDavid102) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:375]: Of course

**Comment by mpellicer** [comment]
Is this ready @JuanDavid102 ?

**Comment by JuanDavid102** [comment]
For me it is ok. But I think that Paco has not finished the review

**Review by mpellicer — APPROVED** [review]

**Comment by mpellicer** [comment]
Let's give @frasese and the community more time to review this.

**Comment by jesusmmp** [comment]
I'm going to test this patch in our environment next week!. Great work!

**Review by mpellicer — CHANGES_REQUESTED** [review]
Changes requested via Jira.

**Comment by mpellicer** [comment]
This has conflicts to be resolved.

**Comment by JuanDavid102** [comment]
All conflicts resolved

**Comment by frasese** [comment]
Comment about finally block should be revised

**Comment by frasese** [comment]
apart from that, looks good to me

**Comment by mpellicer** [comment]
Is this ready to be merged?

**Review by mpellicer — APPROVED** [review]

**Comment by mpellicer** [comment]
Thanks for all the hard work @JuanDavid102 @frasese and @jesusmmp

**Pull request merged**